File: src/lib/activity-ingest.ts
import { desc, eq } from "drizzle-orm";
import { db } from "@/db";
import { activitySegments, userPreferences } from "@/db/schema";
import type {
  ActivityKind,
  ActivitySegmentInput,
} from "@/lib/activity-types";

function newId() {
  return crypto.randomUUID();
}

function asKind(value: string): ActivityKind {
  return value === "idle" ? "idle" : "app";
}

function parseIso(value: string) {
  const date = new Date(value);
  return Number.isNaN(date.getTime()) ? null : date;
}

function segmentKey(row: {
  kind: string;
  processName?: string | null;
  exePath?: string | null;
}) {
  if (row.kind === "idle") return "idle";
  return `${row.exePath ?? ""}|${row.processName ?? ""}`;
}

/** Merge contiguous same-app / idle slices so we don't persist tiny fragments. */
export function coalesceSegments(
  segments: ActivitySegmentInput[],
): ActivitySegmentInput[] {
  const sorted = [...segments].sort(
    (a, b) =>
      new Date(a.startedAt).getTime() - new Date(b.startedAt).getTime(),
  );
  const merged: ActivitySegmentInput[] = [];

  for (const segment of sorted) {
    const startedAt = parseIso(segment.startedAt);
    const endedAt = parseIso(segment.endedAt);
    if (!startedAt || !endedAt || endedAt <= startedAt) continue;

    const kind = asKind(segment.kind);
    const next: ActivitySegmentInput = {
      id: segment.id?.trim() || newId(),
      startedAt: startedAt.toISOString(),
      endedAt: endedAt.toISOString(),
      kind,
      processName: kind === "app" ? (segment.processName ?? null) : null,
      appName: kind === "app" ? (segment.appName ?? null) : null,
      exePath: kind === "app" ? (segment.exePath ?? null) : null,
    };

    const prev = merged[merged.length - 1];
    if (
      prev &&
      segmentKey(prev) === segmentKey(next) &&
      new Date(next.startedAt).getTime() - new Date(prev.endedAt).getTime() <=
        5_000
    ) {
      prev.endedAt =
        new Date(next.endedAt) > new Date(prev.endedAt)
          ? next.endedAt
          : prev.endedAt;
      if (!prev.appName && next.appName) prev.appName = next.appName;
      continue;
    }

    merged.push(next);
  }

  return merged;
}

export async function getActivityTrackingEnabled(
  userId: string,
): Promise<boolean> {
  const row = await db.query.userPreferences.findFirst({
    where: eq(userPreferences.userId, userId),
    columns: { activityTrackingEnabled: true },
  });
  return row?.activityTrackingEnabled ?? false;
}

/**
 * Persist coalesced segments for a user. Extends the latest matching row
 * when the new slice continues the same app/idle streak.
 */
export async function ingestActivitySegmentsForUser(
  userId: string,
  segments: ActivitySegmentInput[],
) {
  if (!Array.isArray(segments) || segments.length === 0) {
    return { inserted: 0, updated: 0 };
  }

  const coalesced = coalesceSegments(segments.slice(0, 200));
  if (coalesced.length === 0) {
    return { inserted: 0, updated: 0 };
  }

  const latest = await db
    .select()
    .from(activitySegments)
    .where(eq(activitySegments.userId, userId))
    .orderBy(desc(activitySegments.endedAt))
    .limit(1);

  const rows = [];
  let updated = 0;
  let cursor = latest[0] ?? null;

  for (const segment of coalesced) {
    const startedAt = parseIso(segment.startedAt)!;
    const endedAt = parseIso(segment.endedAt)!;
    const kind = asKind(segment.kind);
    const segmentId = segment.id?.trim() || newId();

    // Same open-session id → always extend (live flush snapshots).
    if (cursor && cursor.id === segmentId) {
      if (endedAt > cursor.endedAt) {
        await db
          .update(activitySegments)
          .set({
            endedAt,
            appName:
              kind === "app"
                ? (segment.appName ?? cursor.appName)
                : cursor.appName,
          })
          .where(eq(activitySegments.id, cursor.id));
        cursor = {
          ...cursor,
          endedAt,
          appName: segment.appName ?? cursor.appName,
        };
        updated += 1;
      }
      continue;
    }

    if (
      cursor &&
      segmentKey(cursor) ===
        segmentKey({
          kind,
          processName: segment.processName,
          exePath: segment.exePath,
        }) &&
      startedAt.getTime() - cursor.endedAt.getTime() <= 5_000
    ) {
      if (endedAt > cursor.endedAt) {
        await db
          .update(activitySegments)
          .set({
            endedAt,
            appName:
              kind === "app"
                ? (segment.appName ?? cursor.appName)
                : cursor.appName,
          })
          .where(eq(activitySegments.id, cursor.id));
        cursor = {
          ...cursor,
          endedAt,
          appName: segment.appName ?? cursor.appName,
        };
        updated += 1;
      }
      continue;
    }

    const row = {
      id: segmentId,
      userId,
      startedAt,
      endedAt,
      kind,
      processName:
        kind === "app" ? (segment.processName?.slice(0, 256) ?? null) : null,
      appName: kind === "app" ? (segment.appName?.slice(0, 512) ?? null) : null,
      exePath: kind === "app" ? (segment.exePath?.slice(0, 1024) ?? null) : null,
    };
    rows.push(row);
    cursor = {
      ...row,
      createdAt: new Date(),
    };
  }

  if (rows.length > 0) {
    for (const row of rows) {
      await db
        .insert(activitySegments)
        .values(row)
        .onConflictDoUpdate({
          target: activitySegments.id,
          set: {
            endedAt: row.endedAt,
            appName: row.appName,
          },
        });
    }
  }

  return { inserted: rows.length, updated };
}

File: apps/api/app/domain/activity_ingest.py
"""Coalesce and persist activity segments, ported from src/lib/activity-ingest.ts."""

from datetime import datetime, timezone
from uuid import uuid4

from sqlalchemy import desc
from sqlalchemy.orm import Session as DbSession

from app.models import ActivitySegment


def _as_kind(v: str) -> str:
    return "idle" if v == "idle" else "app"


def _as_utc_naive(dt: datetime) -> datetime:
    """DB columns are naive UTC; strip tzinfo after converting to UTC."""
    if dt.tzinfo is None:
        return dt
    return dt.astimezone(timezone.utc).replace(tzinfo=None)


def _parse_iso(v: str) -> datetime | None:
    try:
        parsed = datetime.fromisoformat(v.replace("Z", "+00:00"))
        return _as_utc_naive(parsed)
    except (ValueError, AttributeError):
        return None


def _db_utc_naive(dt: datetime | None) -> datetime | None:
    if dt is None:
        return None
    return _as_utc_naive(dt)


def _seg_key(kind: str, process_name: str | None, exe_path: str | None) -> str:
    if kind == "idle":
        return "idle"
    return f"{exe_path or ''}|{process_name or ''}"


def coalesce(segments: list[dict]) -> list[dict]:
    raw = sorted(segments, key=lambda s: s.get("startedAt", ""))
    merged: list[dict] = []
    for seg in raw:
        started = _parse_iso(seg.get("startedAt", ""))
        ended = _parse_iso(seg.get("endedAt", ""))
        if not started or not ended or ended <= started:
            continue
        kind = _as_kind(seg.get("kind", "app"))
        entry = {
            "id": (seg.get("id") or "").strip() or uuid4().hex,
            "startedAt": started,
            "endedAt": ended,
            "kind": kind,
            "processName": seg.get("processName") if kind == "app" else None,
            "appName": seg.get("appName") if kind == "app" else None,
            "exePath": seg.get("exePath") if kind == "app" else None,
        }
        prev = merged[-1] if merged else None
        if (
            prev
            and _seg_key(prev["kind"], prev.get("processName"), prev.get("exePath"))
            == _seg_key(kind, entry.get("processName"), entry.get("exePath"))
            and (entry["startedAt"] - prev["endedAt"]).total_seconds() <= 5
        ):
            if entry["endedAt"] > prev["endedAt"]:
                prev["endedAt"] = entry["endedAt"]
            if not prev.get("appName") and entry.get("appName"):
                prev["appName"] = entry["appName"]
            continue
        merged.append(entry)
    return merged


def ingest_for_user(db: DbSession, user_id: str, segments: list[dict]) -> dict:
    if not segments:
        return {"inserted": 0, "updated": 0}

    coalesced = coalesce(segments[:200])
    if not coalesced:
        return {"inserted": 0, "updated": 0}

    latest = (
        db.query(ActivitySegment)
        .filter(ActivitySegment.user_id == user_id)
        .order_by(desc(ActivitySegment.ended_at))
        .first()
    )

    rows: list[dict] = []
    updated = 0
    cursor = latest

    for seg in coalesced:
        started_at = seg["startedAt"]
        ended_at = seg["endedAt"]
        kind = seg["kind"]
        seg_id = seg["id"]
        cursor_ended = _db_utc_naive(cursor.ended_at) if cursor else None

        if cursor and cursor.id == seg_id:
            if cursor_ended is not None and ended_at > cursor_ended:
                cursor.ended_at = ended_at
                if kind == "app" and seg.get("appName"):
                    cursor.app_name = seg["appName"]
                updated += 1
            continue

        if (
            cursor
            and cursor_ended is not None
            and _seg_key(cursor.kind, cursor.process_name, cursor.exe_path)
            == _seg_key(kind, seg.get("processName"), seg.get("exePath"))
            and (started_at - cursor_ended).total_seconds() <= 5
        ):
            if ended_at > cursor_ended:
                cursor.ended_at = ended_at
                if kind == "app" and seg.get("appName"):
                    cursor.app_name = seg["appName"]
                updated += 1
            continue

        row = ActivitySegment(
            id=seg_id,
            user_id=user_id,
            started_at=started_at,
            ended_at=ended_at,
            kind=kind,
            process_name=(seg.get("processName") or "")[:256] or None if kind == "app" else None,
            app_name=(seg.get("appName") or "")[:512] or None if kind == "app" else None,
            exe_path=(seg.get("exePath") or "")[:1024] or None if kind == "app" else None,
        )
        db.merge(row)
        rows.append(row)
        cursor = row

    return {"inserted": len(rows), "updated": updated}
